src/log-store/src/raft_engine/backend.rs

```diff
@@ -0,0 +1,574 @@
+        }
+
+        engine.write(&mut batch, false).context(RaftEngineSnafu)?;
+
+        Ok(BatchPutResponse { prev_kvs })
+    }
+
+    async fn batch_get(&self, req: BatchGetRequest) -> Result<BatchGetResponse, Self::Error> {
+        let mut response = BatchGetResponse {
+            kvs: Vec::with_capacity(req.keys.len()),
+        };
+        let engine = self.engine.read().unwrap();
+        for key in req.keys {
+            let Some(value) = engine.get(SYSTEM_NAMESPACE, &key) else {
+                continue;
+            };
+            response.kvs.push(KeyValue { key, value });
+        }
+        Ok(response)
+    }
+
+    async fn compare_and_put(
+        &self,
+        req: CompareAndPutRequest,
+    ) -> Result<CompareAndPutResponse, Self::Error> {
+        let CompareAndPutRequest { key, expect, value } = req;
+
+        let mut batch = LogBatch::with_capacity(1);
+        let engine = self.engine.write().unwrap();
+        let existing = engine_get(&engine, &key)?;
+        let eq = existing
+            .as_ref()
+            .map(|kv| kv.value == expect)
+            .unwrap_or_else(|| {
+                // if the associated value of key does not exist and expect is empty,
+                // then we still consider them as equal.
+                expect.is_empty()
+            });
+
+        if eq {
+            batch
+                .put(SYSTEM_NAMESPACE, key, value)
+                .context(RaftEngineSnafu)?;
+            engine.write(&mut batch, false).context(RaftEngineSnafu)?;
+        }
+        Ok(CompareAndPutResponse {
+            success: eq,
+            prev_kv: existing,
+        })
+    }
+
+    async fn delete_range(
+        &self,
+        req: DeleteRangeRequest,
+    ) -> Result<DeleteRangeResponse, Self::Error> {
+        let DeleteRangeRequest {
+            key,
+            range_end,
+            prev_kv,
+        } = req;
+
+        let range = RangeRequest {
+            key,
+            range_end,
+            limit: 0,
+            keys_only: true,
+        };
+        let range_resp = self.range(range).await?;
+
+        let mut prev_kvs = vec![];
+        let mut deleted = 0;
+
+        let engine = self.engine.read().unwrap();
+        for kv in range_resp.kvs {
+            engine_delete(&engine, &kv.key)?;
+            if prev_kv {
+                prev_kvs.push(kv);
+            }
+            deleted += 1;
+        }
+
+        Ok(DeleteRangeResponse { deleted, prev_kvs })
+    }
+
+    async fn batch_delete(
+        &self,
+        req: BatchDeleteRequest,
+    ) -> Result<BatchDeleteResponse, Self::Error> {
+        let BatchDeleteRequest { keys, prev_kv } = req;
+
+        let mut prev_kvs = if prev_kv {
+            Vec::with_capacity(keys.len())
+        } else {
+            vec![]
+        };
+        let mut batch = LogBatch::with_capacity(keys.len());
+
+        for key in keys {
+            if prev_kv && let Some(prev) = engine_get(&self.engine.read().unwrap(), &key)? {
```

Comment:
Should this read lock acquisition be lifted out of for loop as well just like `batch_get` does?